Egui editor

vaneri-9 opened the pull request:
(Moved from #19 accidentally used wrong remote)
Related #17

Todo (not all of these are in scope for this pr):
 - [x] Rename `InputId` to `InputPort` and the `id` field to `port_id` and all variables using it
 - [x] Refactor/rename functions and where functionality is
 - [x] `Save`, `Save as`, `Load`
 - [x] Fix `mem` tests
 - [x] Remove comments on line 57 in `common.rs`
 - [x] Reverse Serialize/Deserialize on line 76 in `common.rs` (to revert)
 - [x] Add a button to enter wire mode
 - [ ] Care about different signal types when connecting input (`Combinatorial` etc)
 - [x] Auto-save when switching from editor in case you did something wrong
 - [ ] Select multiple components with a drag and select feature (BIG)
 - [ ] Properties maybe should be on the left side (where the library is)
 - [x] Properties window should not include sliders for any number input, there are issues with strobing and they don't really make sense for what they're supposed to do
 - [x] Snapping for components to the grid when moving them (maybe also placing)
 - [ ] Add more components
   - [X] add
   - [X] constant
   - [x] mem
   - [x] mux
   - [X] probe
   - [x] probe_edit
   - [ ] probe_out
   - [ ] probe_stim
   - [ ] probe_assert
   - [x] register
   - [x] sext
   - [X] wire
 
Changes made:
- Simulator now consumes the `ComponentStore`, this is required since we can only have ONE reference to a `dyn Component` to be able to edit it because of `Rc` limitation (it is intended to work this way)
- Every component requires to be made with the `::new` method since otherwise we don't conditionally create the `EguiExtra` struct that egui needs for editor functionality
- The `InputPort` wraps an `Input` but also provides an id for a port, `gui-egui` requires this to bind the ends of two wires
- `gui-egui` now also has a `render_editor` function trait that requires the `dyn Component` to be mutable and only runs during a specific `EditorMode` in the editor.
- `gui-egui` render function returns `Option<Vec<egui::Response>>` because of the `wire` being multiple responses (the points)
- `gui-egui` components now have tooltips (some are broken)
 
 Bugs:
- [x] Program crashes when starting simulator with an incorrect `ComponentStore`, there is auto-save.
- [ ] Moving wires does not connect or disconnect anything, after they've been placed they are purely visual
- [x] You can not start or end at a wire atm. Functionality will be added in the future (reason for it not existing now is that if a wire is connected to a port and you're trying to also connect to that port the wire should prefer connecting to the port and not the wire, this functionality does not exist atm)
- [x] Properties window closes when you select an item that was far down on the dropdown (outside the properties window)
- [x] Switching to editor, simulation, editor will crash on last switch
- [x] Right-clicking a multi-position wire will open the window and quickly close it because of `clicked_elsewhere` working on the individual part of the segments
- [ ] When changing the id of a component also somehow change all the items connected to that component to update to the new id
- [ ] Editor mode doesn't have background colors for `Constant` and `Probe` to be able to show where the ports are located
- [ ] Moving your mouse outside the window while placing a component will make the cursor offset from the component when reentering
- [x] Placing a component while being in scaling doesn't place it in the correct position
- [x] Delete component with keyboard or button
- [x] Loading a model does not work (anymore)
- [ ] Right clicking in input mode on the library crashes the program
- [ ] Holding left click selecting components and then pressing right-click panics the program 

How to use:
- Toggle editor `Ctrl+e` or `File`->`Editor`
- Save: `Ctrl+s` (currently only saves to `file.json`)
- Left side is the library of components, click a component entering place mode for it where you can place it until you press `Escape` (or `Right-click`
- `w` to enter wire mode so you can connect multiple components with wires, it will automatically snap to open ports on the components
- Right-click components to manually edit them
- Left-click-drag components to move them around
 
I've been testing with `cargo run --no-default-features --features=gui-egui --example add -- --model mips`

vaneri-9 commented:
Finally ci passes!

onsdagens commented:
Bugs found (in no particular order, i'll update as i go):

Using the slider to set a Constant value past a certain threshold causes intense strobing.

Clicking the Editor option in the File Menu causes panic
```  thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/gui_egui/menu.rs:35:66 ``` 

In wire mode, connecting a wire to it's point of origin causes panic
``` thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/gui_egui/editor_wire_mode.rs:83:22 ```

vaneri-9 commented:
> Clicking the Editor option in the File Menu causes panic `` thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/gui_egui/menu.rs:35:66 ``

Fixed
> In wire mode, connecting a wire to it's point of origin causes panic `` thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/gui_egui/editor_wire_mode.rs:83:22 ``

Fixed
> Suggestions:
> 
> Since wires are just visual, listing them in the input selection dropdown seems unnecessary, skipping them would save a lot of clutter in the list

Fixed

I haven't fixed the strobing yet, I'm going to keep it out of scope of this pr but add it to the todo on the original comment.

perlindgren commented:
A few suggestions:

- Grid, with settable size, and snap to grid for component placement

onsdagens commented on src/gui_egui/editor.rs:
```diff
@@ -0,0 +1,405 @@
+impl Editor {
+    pub fn gui(components: Components, _path: &Path) -> Self {
+        let dummy_input = Input::new("id", "field");
+        let library: Components = vec![
```
this should probably be passed as an argument, allowing model crates (like riscv) to provide their own component libraries

onsdagens reviewed (commented):
I'm using this now to create a better model view for the riscv, comments incoming...

onsdagens commented:
Found a bug, the left side library view will not scroll when there are more elements than fit the screen. I've fixed it for now by just decreasing the render scale.

onsdagens commented:
This is now partly merged with #78 . Since the feature is still experimental (i.e. buggy), let's keep this issue around for tracking.
